File: web/src/lib/timesgate.ts
// Times Gate device command builders (mirror of timesgate.py).
// Pure payload construction — transport is handled by bridge.ts.
//
// Firmware rules encoded here (see repo README for the full story):
//  - every command carries an integer LocalToken
//  - Draw/SendHttpGif PicData = base64 JPEG (not raw RGB)
//  - PicID = small strictly-increasing counter (reset via Draw/ResetHttpGifId)
//  - drawing selects screens with LcdArray[5]; text/items use LcdIndex

export const SCREEN_COUNT = 5;
export const IMG_SIZE = 128;

export type Command = Record<string, unknown>;

// Shared monotonic PicID counter (device rejects reused/lower ids).
let picId = 1;
export function nextPicId(): number {
  return picId++;
}
export function resetPicIdCounter() {
  picId = 1;
}

/** int or list of 0-based indices -> length-5 0/1 mask. */
export function lcdMask(screens: number | number[]): number[] {
  const list = typeof screens === "number" ? [screens] : screens;
  const mask = Array(SCREEN_COUNT).fill(0);
  for (const s of list) mask[s] = 1;
  return mask;
}

/** First active screen in a mask (the LcdIndex a text overlay must target). */
export function firstActive(mask: number[]): number {
  const i = mask.findIndex((v) => v === 1);
  return i < 0 ? 0 : i;
}

/** Encode a canvas as base64 JPEG (Times Gate wants JPEG, quality ~0.95). */
export async function canvasToJpegBase64(
  canvas: HTMLCanvasElement,
  quality = 0.95,
): Promise<string> {
  const blob: Blob = await new Promise((resolve, reject) =>
    canvas.toBlob(
      (b) => (b ? resolve(b) : reject(new Error("toBlob failed"))),
      "image/jpeg",
      quality,
    ),
  );
  const bytes = new Uint8Array(await blob.arrayBuffer());
  let binary = "";
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  return btoa(binary);
}

export const resetHttpGifId = (): Command => ({ Command: "Draw/ResetHttpGifId" });

/** One Draw/SendHttpGif packet (one frame). */
export function sendHttpGifFrame(opts: {
  mask: number[];
  picNum: number;
  picOffset: number;
  picId: number;
  picSpeed: number;
  picData: string;
}): Command {
  return {
    Command: "Draw/SendHttpGif",
    LcdArray: opts.mask,
    PicNum: opts.picNum,
    PicWidth: IMG_SIZE,
    PicOffset: opts.picOffset,
    PicID: opts.picId,
    PicSpeed: opts.picSpeed,
    PicData: opts.picData,
  };
}

/** Build the full packet list for an animation (share one PicID across frames). */
export function buildAnimation(
  screens: number | number[],
  frames: string[], // base64 JPEG per frame
  speedMs: number,
): Command[] {
  const mask = lcdMask(screens);
  const id = nextPicId();
  return frames.map((picData, picOffset) =>
    sendHttpGifFrame({
      mask,
      picNum: frames.length,
      picOffset,
      picId: id,
      picSpeed: speedMs,
      picData,
    }),
  );
}

/** Text overlay. Must run AFTER a SendHttpGif on that screen. TextWidth 17..63. */
export function sendHttpText(opts: {
  lcdIndex: number;
  text: string;
  x?: number;
  y?: number;
  color?: string;
  font?: number;
  width?: number;
  speed?: number;
  dir?: 0 | 1;
  textId?: number;
  align?: 1 | 3 | 5;
}): Command {
  const width = Math.max(17, Math.min(opts.width ?? 56, 63));
  return {
    Command: "Draw/SendHttpText",
    // Firmware is inconsistent about the field name for the target screen:
    // API docs say LcdIndex, but working apps (adiastra) use LcdId. Send both.
    LcdIndex: opts.lcdIndex,
    LcdId: opts.lcdIndex,
    TextId: opts.textId ?? 1,
    x: opts.x ?? 0,
    y: opts.y ?? 52,
    dir: opts.dir ?? 0,
    font: opts.font ?? 4,
    TextWidth: width,
    speed: opts.speed ?? 60,
    TextString: opts.text,
    color: opts.color ?? "#FFFFFF",
    align: opts.align ?? 1,
  };
}

export type ItemListItem = {
  TextId: number;
  type: number; // 1-21 on-device data, 22 static, 23 net-text(URL)
  x: number;
  y: number;
  dir?: 0 | 1;
  font: number;
  TextWidth: number;
  Textheight?: number;
  TextString?: string;
  speed?: number;
  color?: string;
  update_time?: number;
  align?: number; // firmware uses 0-based here (0/1/2), not the 1/3/5 of SendHttpText
};

/** Self-updating item list. NewFlag 1 = set background + items; BackgroudGif URL required then. */
export function sendHttpItemList(opts: {
  lcdIndex: number;
  items: ItemListItem[];
  newFlag?: 0 | 1;
  backgroundGif?: string;
}): Command {
  const cmd: Command = {
    Command: "Draw/SendHttpItemList",
    LcdIndex: opts.lcdIndex,
    NewFlag: opts.newFlag ?? 1,
    ItemList: opts.items,
  };
  if (opts.backgroundGif) cmd.BackgroudGif = opts.backgroundGif;
  return cmd;
}

// Hosted solid background gifs (device fetches these; palette index 0 is an
// unused sentinel so the solid colour isn't treated as transparent).
export const BG_BASE = "https://pixelgate.pages.dev/bg/solid-";
// Pages Function that echoes ?t=<text> as {"DispData": text} for type-23 net-text.
export const ECHO_BASE = "https://pixelgate.pages.dev/api/echo";
export type BgPreset = "dark" | "black" | "navy" | "plum" | "white";

/**
 * Scrolling text / marquee via on-device SendHttpItemList type 23 (net-text).
 * The device polls our echo Function for the text and scrolls it when it
 * overflows TextWidth. font 2 (has letters) at full width.
 */
export function buildScrollingText(
  screens: number | number[],
  opts: { text: string; color: string; bg: BgPreset; speed?: number; y?: number },
): Command[] {
  const list = typeof screens === "number" ? [screens] : screens;
  const bgUrl = `${BG_BASE}${opts.bg}.gif`;
  const url = `${ECHO_BASE}?t=${encodeURIComponent(opts.text)}`;
  return list.map((s) =>
    sendHttpItemList({
      lcdIndex: s,
      newFlag: 1,
      backgroundGif: bgUrl,
      items: [
        {
          TextId: 1,
          type: 23, // net-text: polls the URL, scrolls on overflow
          x: 0,
          y: opts.y ?? 52,
          dir: 0,
          font: 2, // general font with letters (numeral fonts have no A-Z)
          TextWidth: 128, // full-width scroll region
          Textheight: 16,
          speed: opts.speed ?? 30,
          color: opts.color,
          update_time: 60,
          TextString: url,
        },
      ],
    }),
  );
}

// Fonts that actually render on this firmware. Big built-in fonts (242/256/260…)
// silently blank, so the "large" option tops out at font 90.
const CLOCK_FONT = { small: 2, large: 90 } as const;

/**
 * On-device self-updating digital clock (ticks natively, no re-push).
 * HH:MM on top; optional SS stacked smaller below. Position/size/colour are
 * tunable because centring/align is unreliable on this firmware.
 */
export function buildDigitalClock(
  screens: number | number[],
  opts: {
    color: string;
    big: boolean;
    seconds: boolean;
    x: number;
    y: number;
    bg: BgPreset;
  },
): Command[] {
  const list = typeof screens === "number" ? [screens] : screens;
  const bgUrl = `${BG_BASE}${opts.bg}.gif`;
  const font = opts.big ? CLOCK_FONT.large : CLOCK_FONT.small;
  return list.map((s) => {
    const items: ItemListItem[] = [
      {
        TextId: 1,
        type: 5, // HH:MM
        x: opts.x,
        y: opts.y,
        dir: 0,
        font,
        TextWidth: 128,
        Textheight: 16,
        speed: 100,
        color: opts.color,
        update_time: 1,
        align: 0,
      },
    ];
    if (opts.seconds) {
      items.push({
        TextId: 2,
        type: 1, // seconds
        x: opts.x + 17,
        y: opts.y + 30,
        dir: 0,
        font: 2,
        TextWidth: 60,
        Textheight: 16,
        speed: 100,
        color: opts.color,
        update_time: 1,
        align: 0,
      });
    }
    return sendHttpItemList({ lcdIndex: s, newFlag: 1, backgroundGif: bgUrl, items });
  });
}

File: timesgate.py
"""Minimal Divoom Times Gate client (5x 128x128 LCD screens, Hardware 400).

Firmware gotchas (learned the hard way — see README):
  * EVERY /post request must carry an integer ``LocalToken`` (shown in the Divoom
    phone app). Without it every command returns ``{"error_code": "DeviceToken is err"}``.
  * ``Draw/SendHttpGif`` ``PicData`` must be base64 **JPEG** (q~95), NOT raw RGB.
    Raw RGB returns error_code 0 but leaves the screen stuck on "loading".
  * ``PicID`` must be a SMALL, strictly-increasing counter (1,2,3,...). Using a
    unix timestamp -> stuck "loading". Reusing/lowering an id -> send ignored.
    Call ``Draw/ResetHttpGifId`` once at startup, then use the shared counter.
  * Drawing commands select screens with ``LcdArray`` (length-5 0/1 mask).
    Per-screen commands (text) use ``LcdIndex`` (single int 0-4).
"""

import base64
import io
import itertools

import requests
from PIL import Image

SCREEN_COUNT = 5
IMG_SIZE = 128


class TimesGate:
    def __init__(self, ip, local_token, timeout=8):
        self.ip = ip.strip()
        self.local_token = int(local_token)
        self.timeout = timeout
        self.url = f"http://{self.ip}/post"
        # shared, monotonic PicID counter across all screens
        self._pic_id = itertools.count(1)

    # --- low level -------------------------------------------------------
    def post(self, command, retries=2):
        """POST with LocalToken injected. Retries transient timeouts — the device
        drops its HTTP server briefly under load / WiFi power-save."""
        payload = {**command, "LocalToken": self.local_token}
        for attempt in range(retries + 1):
            try:
                r = requests.post(self.url, json=payload, timeout=self.timeout)
                r.raise_for_status()
                try:
                    return r.json()
                except ValueError:
                    return {"raw": r.text}
            except (requests.Timeout, requests.ConnectionError):
                if attempt == retries:
                    raise

    def check(self):
        """True if the device accepts our LocalToken."""
        resp = self.post({"Command": "Device/GetDeviceTime"})
        return "error_code" not in resp or resp.get("error_code") in (0, "0")

    def reset_gif_id(self):
        """Reset device PicID counter; call once before a fresh drawing session."""
        self._pic_id = itertools.count(1)
        return self.post({"Command": "Draw/ResetHttpGifId"})

    # --- helpers ---------------------------------------------------------
    @staticmethod
    def lcd_mask(screens):
        """int or iterable of 0-based indices -> length-5 0/1 mask."""
        if isinstance(screens, int):
            screens = [screens]
        mask = [0] * SCREEN_COUNT
        for s in screens:
            mask[s] = 1
        return mask

    @staticmethod
    def _encode(img, quality=95):
        img = img.convert("RGB")
        if img.size != (IMG_SIZE, IMG_SIZE):
            img = img.resize((IMG_SIZE, IMG_SIZE))
        buf = io.BytesIO()
        img.save(buf, format="JPEG", quality=quality)
        return base64.b64encode(buf.getvalue()).decode()

    # --- drawing ---------------------------------------------------------
    def send_frames(self, screens, frames, speed=100, quality=95):
        """Push PIL.Image frames to screen(s). speed = ms/frame. Max 59 frames."""
        if not frames:
            raise ValueError("no frames")
        if len(frames) >= 60:
            raise ValueError("PicNum must be < 60")
        mask = self.lcd_mask(screens)
        pic_id = next(self._pic_id)
        n = len(frames)
        last = None
        for offset, frame in enumerate(frames):
            last = self.post({
                "Command": "Draw/SendHttpGif",
                "LcdArray": mask,
                "PicNum": n,
                "PicWidth": IMG_SIZE,
                "PicOffset": offset,
                "PicID": pic_id,
                "PicSpeed": speed,
                "PicData": self._encode(frame, quality),
            })
        return last

    def send_image(self, screens, img, quality=95):
        return self.send_frames(screens, [img], quality=quality)

    def send_text(self, screen_index, text, x=0, y=52, color="#FFFFFF",
                  font=4, width=56, speed=60, direction=0, text_id=1, align=1):
        """Overlay text on a screen. Must run AFTER a SendHttpGif on that screen.
        Uses LcdIndex (single int) = the screen that already has a gif.
        TextWidth must be >16 and <64 (>=64 -> "Request data illegal json").
        Text scrolls automatically when longer than TextWidth. font: even ids
        only (2,4,18,20,...); align 1=left 3=middle 5=right; TextId unique <20."""
        width = max(17, min(int(width), 63))
        return self.post({
            "Command": "Draw/SendHttpText",
            "LcdIndex": int(screen_index),
            "TextId": text_id,
            "x": x, "y": y,
            "dir": direction,          # 0 = scroll left, 1 = scroll right
            "font": font,
            "TextWidth": width,
            "speed": speed,
            "TextString": text,
            "color": color,
            "align": align,
        })


if __name__ == "__main__":
    import sys
    if len(sys.argv) < 3:
        sys.exit("usage: python timesgate.py <ip> <LocalToken> [screen]")
    ip, token = sys.argv[1], sys.argv[2]
    screen = int(sys.argv[3]) if len(sys.argv) > 3 else 0

    gate = TimesGate(ip, token)
    print("token accepted:", gate.check())
    print("reset id:", gate.reset_gif_id())

    img = Image.new("RGB", (IMG_SIZE, IMG_SIZE), (200, 30, 30))
    print("sending solid red to screen", screen, "->", gate.send_image(screen, img))
